### PR TITLE
implement lxd relation

### PR DESCRIPTION
**Pull request by aguadoenzo**
This new 'LXD' relation is an early draft meant to fill the gap between
any LXD providing charm and their clients. This integrator charm makes use
of that relation
Included is the operators framework implementation of the relation

### CODE REVIEW

File: interface/README.md [hunk 1]
```diff
@@ -0,0 +1,51 @@
+# LXD Interface
```
Comment (morphis) [inline interface/README.md]: As discussed we don't want the `interface` directory but just the provider side and that directly in `src/interfaces.py` similar to how other operator charms are doing it.

File: interface/README.md [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+The opposite role would then of course have to read data on the relation coming
+from the other side and work with it.
+
+Data for each key should be published as JSON.
+
+#### Provider
+```yaml
+"nodes": [
+    "https://10.10.10.10:8443",
+    "https://10.10.10.11:8444"
```
Comment (morphis) [inline interface/README.md]: Why do we have different ports? LXD usually always listens on 8443

File: interface/README.md [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+],
+"registered_certificates": [
```
Comment (morphis) [inline interface/README.md]: Why do we provide a list of registered certificates?

File: interface/metadata.yaml [hunk 4]
```diff
@@ -0,0 +1,5 @@
+name: lxd
```
Comment (morphis) [inline interface/metadata.yaml]: A metadata.yaml doesn't exist for an interface in the operator framework world. It's just a python library, nothing else.

File: interface/provides.py [hunk 5]
```diff
@@ -0,0 +1,196 @@
+#!/usr/bin/env python3
+#
+# (c) 2020 Canonical Ltd. All right reserved
```
Comment (morphis) [inline interface/provides.py]: right -> rights

File: interface/requires.py [hunk 6]
```diff
@@ -0,0 +1,56 @@
+#!/usr/bin/env python3
+#
+# (c) 2020 Canonical Ltd. All right reserved
```
Comment (morphis) [inline interface/requires.py]: right -> rights

File: src/charm.py [hunk 7]
```diff
@@ -1,15 +1,20 @@
 #!/usr/bin/env python3
 #
-# (c) 2020 Canonical Ltd. All right reservered
+# (c) 2020 Canonical Ltd. All right reserved
```
Comment (morphis) [inline src/charm.py]: right -> rights

File: src/charm.py [hunk 8]
```diff
@@ -1,15 +1,20 @@
...
+import subprocess
+
 from ops.charm import CharmBase
 from ops.main import main
 from ops.model import ActiveStatus, BlockedStatus
 
-import logging
-import json
-import subprocess
+sys.path.append('interface')
```
Comment (morphis) [inline src/charm.py]: We don't need this if we just have `src/interfaces.py`

File: interface/provides.py [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+    def is_ready(self) -> bool:
+        return all([self.state.endpoint, self.state.client_cert,
+                    self.state.client_key, self.state.server_cert])
+
+
+    def _new_connection(self) -> httpclient.HTTPSConnection:
+        """
+        Return a http.client.HTTPSConnection configured with the proper endpoint and certificates
+        """
+        self._write_certs_to_filesystem()
```
Comment (morphis) [inline interface/provides.py]: We should not keep the certificates on the filesystem if we don't need them. When we're done lets remove them again.

File: interface/requires.py [hunk 10]
```diff
@@ -0,0 +1,56 @@
+#!/usr/bin/env python3
```
Comment (morphis) [inline interface/requires.py]: As we don't have any user of this yet we can drop this unless we do the split into a separate library for the interface implementation

File: interface/provides.py [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+        self.state.server_cert = server_cert
+
+        # Check that the credentials are trusted to LXD
+        conn = self._new_connection()
+        conn.request('GET', '/1.0')
+        res = conn.getresponse().read().decode('utf-8')
+        conn.close()
+        if json.loads(res)['metadata']['auth'] != 'trusted':
+            raise RuntimeError('invalid credentials: not trusted')
+        logger.info('credentials configured')
```
Comment (morphis) [inline interface/provides.py]: The logging is inconsistent. Please either start with lower case or upper case but not both.

**Review by morphis — CHANGES_REQUESTED** [review]

File: interface/README.md (hunk 3 above)
Comment (aguadoenzo) [reply interface/README.md]: It's used to notify members of the relation so they may know when the certificate has been trusted

File: src/interface.py [hunk 12]
```diff
@@ -0,0 +1,214 @@
...
+    def _on_relation_joined(self, event: RelationJoinedEvent):
+        if not self.is_ready:
+            event.defer()
+            return
+
+        data = self.framework.model.get_relation(self._relation_name).data[self.model.unit]
+        nodes = json.loads(data.get('nodes', '[]'))
+        nodes.append({
+            'endpoint': self.state.endpoint,
+            'name': self.model.unit.name.replace("/", "-"),
```
Comment (morphis) [inline src/interface.py]: We need to ask LXD for its name. That is the one we want to ship over the relation. You can get it for example via

`$ curl --unix-socket /var/snap/lxd/common/lxd/unix.socket s/1.0 | jq -r .metadata.environment.server_name`

File: src/interface.py [hunk 13]
```diff
@@ -0,0 +1,214 @@
...
+        sslcontext = ssl.create_default_context(purpose=ssl.Purpose.CLIENT_AUTH, cafile=SERVER_CERT_PATH)
+        sslcontext.load_cert_chain(certfile=CLIENT_CERT_PATH, keyfile=CLIENT_KEY_PATH)
+        # Depending on how it was initialized, the LXD server cert can be configured
+        # with 127.0.0.1 as its CN, failing the verification
+        sslcontext.check_hostname = False
+        return httpclient.HTTPSConnection(self.state.endpoint, context=sslcontext, timeout=5)
+
+
+    def set_credentials(self, endpoint: str, client_cert: str, client_key: str, server_cert: str) -> None:
+        self.state.endpoint = endpoint.replace('https://', '').replace('http://', '')
```
Comment (morphis) [inline src/interface.py]: Why do we replace https:// and http://? That amc doesn't want it doesn't mean the integrator should not provide it. We should provide the full URL here and the other side has to trim the URL to whatever it needs.

File: src/interface.py [hunk 14]
```diff
@@ -0,0 +1,214 @@
...
+        self.state.client_cert = client_cert
+        self.state.client_key = client_key
+        self.state.server_cert = server_cert
+
+        # Check that the credentials are trusted to LXD
+        conn = self._new_connection()
+        conn.request('GET', '/1.0')
+        res = conn.getresponse().read().decode('utf-8')
```
Comment (morphis) [inline src/interface.py]: If we have a trusted connection here we can take the result to extract the name of the LXD node

**Review by morphis — CHANGES_REQUESTED** [review]

File: src/interface.py (hunk 13 above)
Comment (aguadoenzo) [reply src/interface.py]: Hm, good point. It wasn't really for amc but for the python http client which doesn't support having the protocol in the uri, will change that

File: src/interface.py [hunk 15]
```diff
@@ -0,0 +1,218 @@
...
+
+        # Check that the credentials are trusted to LXD
+        conn = self._new_connection()
+        conn.request('GET', '/1.0')
+        raw_res = conn.getresponse().read().decode('utf-8')
+        resp = json.loads(raw_res)['metadata']
+        conn.close()
+        if resp['auth'] != 'trusted':
+            raise RuntimeError('invalid credentials: not trusted')
+        self.state.server_name = resp['server_name']
```
Comment (morphis) [inline src/interface.py]: Does this work? According to the API the server_name is in an environment dict:

`$ curl --unix-socket /var/snap/lxd/common/lxd/unix.socket s/1.0 | jq -r .metadata.environment.server_name`

**Review by morphis — CHANGES_REQUESTED** [review]

**Review by morphis — APPROVED** [review]

**Pull request merged**